bridge/tracker/tracker.go

```diff
@@ -0,0 +1,390 @@
+package tracker
+
+import (
+	"context"
+	"encoding/json"
+	"math/big"
+	"os"
+
+	"github.com/0xPolygon/polygon-edge/types"
+
+	"github.com/hashicorp/go-hclog"
+	"github.com/syndtr/goleveldb/leveldb"
+	"github.com/umbracle/go-web3"
+)
+
+const (
+	lastQueriedBlockNumber = "last-block-num" //	db key
+)
+
+//	Tracker represents an event listener that notifies
+//	for each event emitted on the rootchain (ethereum).
+//	Events of interest are defined in rootchain.go
+type Tracker struct {
+	logger hclog.Logger
+
+	// required block confirmations
+	confirmations uint64
+
+	// newHeads subscription channel
+	headerCh chan *ethHeader
+
+	// event channel
+	eventCh chan []byte
+
+	// cancel funcs
+	cancelSubscription, cancelEventTracking context.CancelFunc
+
+	//	rootchain client
+	client *rootchainClient
+
+	//	db for last processed block number
+	db *leveldb.DB
+}
+
+//	NewEventTracker returns a new tracker able to listen for events
+//	at the desired block depth. Events of interest are defined
+//	in rootchain.go
+func NewEventTracker(logger hclog.Logger, confirmations uint64) (*Tracker, error) {
+	tracker := &Tracker{
+		logger:        logger.Named("event_tracker"),
+		confirmations: confirmations,
+		headerCh:      make(chan *ethHeader, 1),
+		eventCh:       make(chan []byte),
+	}
+
+	//	create rootchain client
+	client, err := newRootchainClient(rootchainWS)
+	if err != nil {
+		logger.Error("cannot connect to rootchain", "err", err)
+
+		return nil, err
+	}
+
+	tracker.client = client
+
+	//	load db (last processed block number)
+	db, err := tracker.loadDB()
+	if err != nil {
+		logger.Error("cannot load db", "err", err)
+
+		return nil, err
+	}
+
+	tracker.db = db
+
+	return tracker, nil
+}
+
+//	Start runs the tracker's listening mechanism:
+//	1. startEventTracking - goroutine keeping track of
+//	the last block queried, as well as collecting events
+//	of interest.
+//
+//	2. startSubscription - goroutine responsible for handling
+//	the subscription object. This object is provided by the
+//	client's subscribeNewHeads method that initiates the subscription.
+func (t *Tracker) Start() error {
+	//	create cancellable contexts
+	ctxSubscription, cancelSubscription := context.WithCancel(context.Background())
+	t.cancelSubscription = cancelSubscription
+
+	ctxEventTracking, cancelEventTracking := context.WithCancel(context.Background())
+	t.cancelEventTracking = cancelEventTracking
+
+	//	start event tracking process early
+	go t.startEventTracking(ctxEventTracking)
+
+	//	create newHeads subscription
+	subscription, err := t.client.subscribeNewHeads(t.headerCh)
+	if err != nil {
+		t.logger.Error("cannot subscribe to rootchain", "err", err)
+
+		return err
+	}
+
+	//	start receiving new header events
+	go t.startSubscription(ctxSubscription, subscription)
```

Comment:
This can be a separate method

Reply:
The only reason I didn't group those 2 together is because I so `Start()` can return an error to the caller. Perhaps there isn't a need for this ? If so, I can move `subscribeNewHeads()` into `StartSubscription()`. It all depends on how should we handle errors in this call.

Reply:
Why not instantiate the `subscription` channel when you create the tracker, instead of when you actually start it?

That way you have error handling during creation, and your start method doesn't do 20 things like it does now, it just starts the whole show.